### DISCUSSION TITLE
Changeing the default circle brush tip to other geometric shape freedrawingbrush

### DISCUSSION THREAD

**Discussion by KonMost** [fabricjs/fabric.js#7666] [Q&A]
Hi all,
I am currently working with the fabricjs library and I was able to get the free drawing running. However I could not figure out how to change the circular brush tip to a different shape such as an ellipse, a slanted line or even a texture similar to the brushes used in Photoshop. In the free drawing demos there are examples to add patterns but the shape of the brush is still a circle.
see image:
![image](https://user-images.githubusercontent.com/62653671/153390447-314568b0-7e1e-4d4b-8675-c4382861591e.png)

the one on the left is what I have in fabricjs the ones on the right is what i would like to achive
![photo_2022-02-10_18-51-48](https://user-images.githubusercontent.com/62653671/153467079-150d95e8-9258-4fd1-aff7-5968af4b279b.jpg)

The functionality to draw the line is already available in fabricjs, i just need the drawn object which makes up the line to be different.
How can this be acchieved?

**Comment by ShaMan123** [comment]
There aren't good docs of fabric gist unfortunately
In general it works in 2 steps:
1. drawing is done on the top context
2. once drawing is done (on mouseup) top context is cleared and a path is created and added to canvas

if you want to create a custom brush you need to override both rendering methods of the brush for the drawing step and override the path creation.

take a look at the [source file](https://github.com/fabricjs/fabric.js/blob/master/src/brushes/circle_brush.class.js)

I suggest you read the brush files and the [eraser](https://github.com/fabricjs/fabric.js/blob/master/src/mixins/eraser_brush.mixin.js) to understand how to override and customize stuff

**Reply by KonMost** [reply]
i modified the circle_brush.class.js to a oval.brush.class.js
I changed lines

36ff dot declaration
99ff in onmouseup function

but I am unsure if this would be enough
[oval_brush.class.zip](https://github.com/fabricjs/fabric.js/files/8079576/oval_brush.class.zip)
.

**Reply by ShaMan123** [reply]
Please share a working online example, it should be easy to help

**Reply by KonMost** [reply]
I added this to codepen. It creates an ellipse but there are some issues.
https://codepen.io/KonMost/pen/yLPpzLO

1. the pointer ellipse is different from the one of the actuall line shape stored as array
2. the spacing between the oval shapes is too large which creates gaps instead of a constant line.

i used the circle brush class and changed it accordingly.

**Reply by KonMost** [reply]
i think this is because i based it on the circle brush and the circle brush is not a single circle but several circles with different diameters and alphas. 
I need to base this on the pencil brush which draws a consistent line... trying that now

**Reply by KonMost** [reply]
here is the pencil based brush but i was not able to change the shape yet. 
https://codepen.io/KonMost/pen/zYPpPrm?editors=0111

**Reply by KonMost** [reply]
hmm it is very difficult to comprehend which class its best to use since they are completely different and
seem to rely on calls to other code which sets the geometric shape.
it would be nice to just have a standard class where all the drawing related properties can be modified in a modular way.
shape, size, color,alpha, cursor, path and dot, edit: and the spacing
if anyone can help me out it would be highly appreciated

**Reply by KonMost** [reply]
> I added this to codepen. It creates an ellipse but there are some issues. https://codepen.io/KonMost/pen/yLPpzLO
> 
this is very close, but quite hard to adjust. the spacing is very off
